cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,216 @@
+package cloudwatch
+
+import (
+	"encoding/json"
+	"reflect"
+	"time"
+
+	"github.com/apex/go-apex"
+)
+
+// var (
+// 	detailTypes = map[string]
+// )
+
+// Event represents a CloudWatch Event
+type Event struct {
+	ID         string    `json:"id"`
+	DetailType string    `json:"detail-type"`
+	Source     string    `json:"source"`
+	Account    string    `json:"account"`
+	Time       time.Time `json:"time"`
+	Region     string    `json:"region"`
+	Resources  []string  `json:"resources"`
+	Detail     Detail    `json:"detail"`
+}
+
+// UnmarshalJSON extracts raw data into structured data
+func (e *Event) UnmarshalJSON(b []byte) error {
+	var aux struct {
+		ID         string          `json:"id"`
+		DetailType string          `json:"detail-type"`
+		Source     string          `json:"source"`
+		Account    string          `json:"account"`
+		Time       time.Time       `json:"time"`
+		Region     string          `json:"region"`
+		Resources  []string        `json:"resources"`
+		Detail     json.RawMessage `json:"detail"`
+	}
+
+	if err := json.Unmarshal(b, &aux); err != nil {
+		return err
+	}
+
+	e.ID = aux.ID
+	e.DetailType = aux.DetailType
+	e.Source = aux.Source
+	e.Account = aux.Account
+	e.Time = aux.Time
+	e.Region = aux.Region
+	e.Resources = aux.Resources
+
+	var detail Detail
+	switch aux.Source {
+	case "aws.autoscaling":
+		detail = &AutoScalingGroupDetail{}
+	case "aws.signin":
+		detail = &ConsoleSignInDetail{}
+	case "aws.ec2":
+		detail = &EC2Detail{}
+	default:
+		detail = &EmptyDetail{}
+	}
+
+	if err := json.Unmarshal(aux.Detail, detail); err != nil {
+		return err
+	}
+
+	e.Detail = detail
+
+	return nil
+}
+
+// Detail represents specifics of the event
+type Detail interface {
+	List() []string
+	Get(key string) (interface{}, bool)
+}
```

Comment:
I wonder if instead of getting too opinionated here with interfaces, if we should just leave Details as `json.RawMessage` and provide the other struct(s), making the user unmarshal that part. 

I haven't looked at all the possible events but I think that might be a bit safer. Let me know what you think!